src/Plugin/EntityBrowser/Display/Modal.php

```diff
@@ -216,20 +276,53 @@ public function widgetAjaxCallback(array &$form, FormStateInterface $form_state)
   }
 
   /**
-   * Helper function to return commands to return in AjaxResponse
+   * Helper function to return commands to return in AjaxResponse.
    *
    * @return array
+   *   An array of ajax commands.
    */
   public function getAjaxCommands(FormStateInterface $form_state) {
     $entities = array_map(function(EntityInterface $item) {return [$item->id(), $item->uuid(), $item->getEntityTypeId()];}, $form_state->get(['entity_browser', 'selected_entities']));
+
     $commands = array();
     $commands[] = new SelectEntitiesCommand($this->uuid, $entities);
+    // This doesn't work for now. Probably because we are in the iframe context.
     $commands[] = new CloseDialogCommand();
 
     return $commands;
   }
 
   /**
+   * KernelEvents::RESPONSE listener.
+   *
+   * Intercepts default response and injects
+   * response that will trigger JS to propagate selected entities upstream.
+   *
+   * @param FilterResponseEvent $event
+   *   Response event.
+   */
+  public function propagateSelection(FilterResponseEvent $event) {
```

Comment:
This looks very similar to iFrame display. At some point we should make this class extend that display (if we end up sharing a lot of code). Can be a follow-up.